File: app/src/main/java/com/robifr/ledger/ui/filtercustomer/recycler/FilterCustomerAdapter.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.filtercustomer.recycler;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.robifr.ledger.databinding.CustomerCardWideBinding;
import com.robifr.ledger.databinding.ListableListSelectedItemBinding;
import com.robifr.ledger.ui.RecyclerViewHolder;
import com.robifr.ledger.ui.filtercustomer.FilterCustomerFragment;
import java.util.Arrays;
import java.util.Objects;
import java.util.Optional;

public class FilterCustomerAdapter extends RecyclerView.Adapter<RecyclerViewHolder> {
  private enum ViewType {
    HEADER(0),
    LIST(1);

    private final int _value;

    private ViewType(int value) {
      this._value = value;
    }

    public int value() {
      return this._value;
    }
  }

  @NonNull private final FilterCustomerFragment _fragment;

  public FilterCustomerAdapter(@NonNull FilterCustomerFragment fragment) {
    this._fragment = Objects.requireNonNull(fragment);
  }

  @Override
  @NonNull
  public RecyclerViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    Objects.requireNonNull(parent);

    final ViewType type =
        Arrays.stream(ViewType.values())
            .filter(e -> e.value() == viewType)
            .findFirst()
            .orElse(ViewType.LIST);
    final LayoutInflater inflater = this._fragment.getLayoutInflater();

    return switch (type) {
      case HEADER ->
          new FilterCustomerHeaderHolder(
              this._fragment, ListableListSelectedItemBinding.inflate(inflater, parent, false));

        // Defaults to `ViewType#LIST`.
      default ->
          new FilterCustomerListHolder(
              this._fragment, CustomerCardWideBinding.inflate(inflater, parent, false));
    };
  }

  @Override
  public void onBindViewHolder(@NonNull RecyclerViewHolder holder, int index) {
    Objects.requireNonNull(holder);

    if (holder instanceof FilterCustomerHeaderHolder headerHolder) {
      headerHolder.bind(Optional.empty());

    } else if (holder instanceof FilterCustomerListHolder listHolder) {
      // -1 offset because header holder.
      listHolder.bind(
          this._fragment.filterCustomerViewModel().customers().getValue().get(index - 1));
    }
  }

  @Override
  public int getItemCount() {
    // +1 offset because header holder.
    return this._fragment.filterCustomerViewModel().customers().getValue().size() + 1;
  }

  @Override
  public int getItemViewType(int index) {
    return switch (index) {
      case 0 -> ViewType.HEADER.value();
      default -> ViewType.LIST.value();
    };
  }
}

File: app/src/main/java/com/robifr/ledger/ui/filtercustomer/recycler/FilterCustomerAdapter.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.filtercustomer.recycler

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.robifr.ledger.databinding.CustomerCardWideBinding
import com.robifr.ledger.databinding.ListableListSelectedItemBinding
import com.robifr.ledger.ui.common.RecyclerViewHolder
import com.robifr.ledger.ui.filtercustomer.FilterCustomerFragment

class FilterCustomerAdapter(private val _fragment: FilterCustomerFragment) :
    RecyclerView.Adapter<RecyclerViewHolder>() {
  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerViewHolder =
      when (ViewType.entries.find { it.value == viewType }) {
        ViewType.HEADER ->
            FilterCustomerHeaderHolder(
                _selectedItemBinding =
                    ListableListSelectedItemBinding.inflate(
                        _fragment.layoutInflater, parent, false),
                _filteredCustomers = {
                  _fragment.filterCustomerViewModel.uiState.safeValue.filteredCustomers
                },
                _onCustomerCheckedChanged =
                    _fragment.filterCustomerViewModel::onCustomerCheckedChanged)
        else ->
            FilterCustomerListHolder(
                _cardBinding =
                    CustomerCardWideBinding.inflate(_fragment.layoutInflater, parent, false),
                _customers = {
                  _fragment.filterCustomerViewModel.uiState.safeValue.pagination.paginatedItems
                },
                _onCustomerCheckedChanged =
                    _fragment.filterCustomerViewModel::onCustomerCheckedChanged,
                _filteredCustomers = {
                  _fragment.filterCustomerViewModel.uiState.safeValue.filteredCustomers
                },
                _onExpandedCustomerIndexChanged =
                    _fragment.filterCustomerViewModel::onExpandedCustomerIndexChanged,
                _expandedCustomer = {
                  _fragment.filterCustomerViewModel.uiState.safeValue.expandedCustomer
                })
      }

  override fun onBindViewHolder(holder: RecyclerViewHolder, position: Int) {
    when (holder) {
      is FilterCustomerHeaderHolder -> holder.bind()
      is FilterCustomerListHolder -> holder.bind(position - 1)
    }
  }

  override fun getItemCount(): Int =
      // +1 offset because header holder.
      _fragment.filterCustomerViewModel.uiState.safeValue.pagination.paginatedItems.size + 1

  override fun getItemViewType(position: Int): Int =
      when (position) {
        0 -> ViewType.HEADER.value
        else -> ViewType.LIST.value
      }

  private enum class ViewType(val value: Int) {
    HEADER(0),
    LIST(1)
  }
}
